addons/mail/models/mail_channel.py

```diff
@@ -724,30 +726,51 @@ def _execute_channel_pin(self, pinned=False):
         if channel_partners:
             channel_partners.write({'is_pinned': pinned})
 
-    def channel_seen(self):
+    def channel_seen(self, last_message_id=None):
+        """
+        Mark channel as seen by updating seen message id of the current logged partner
+        :param last_message_id: the id of the message to be marked as seen, last message of the
+        thread by default. This param SHOULD be required, the default behaviour is DEPRECATED and
+        kept only for compatibility reasons.
+        """
         self.ensure_one()
-        if self.channel_message_ids.ids:
-            last_message_id = self.channel_message_ids.ids[0] # zero is the index of the last message
-            channel_partner = self.env['mail.channel.partner'].search([('channel_id', 'in', self.ids), ('partner_id', '=', self.env.user.partner_id.id)], limit=1)
-            if channel_partner.seen_message_id.id == last_message_id:
-                # last message seen by user is already up-to-date
+        domain = [('channel_ids', 'in', self.ids)]
+        if last_message_id:
+            domain = expression.AND([domain, [('id', '<=', last_message_id)]])
+        last_message = self.env['mail.message'].search(domain, order="id DESC", limit=1)
+        if not last_message:
+            return
+
+        channel_partner = self.env['mail.channel.partner'].search([
+            ('channel_id', 'in', self.ids),
+            ('partner_id', '=', self.env.user.partner_id.id)
+        ], limit=1)
+        if channel_partner:
+            if channel_partner.seen_message_id and channel_partner.seen_message_id.id >= last_message.id:
                 return
             channel_partner.write({
-                'seen_message_id': last_message_id,
-                'fetched_message_id': last_message_id,
+                'fetched_message_id': last_message.id,
+                'seen_message_id': last_message.id,
```

Comment:
We're not very strict about alphabetical order in python, don't need to change them usually. It doesn't really matter though so ok since you have diff on the lines anyway.